presto-main/src/main/java/com/facebook/presto/sql/planner/CachingPlanCanonicalInfoProvider.java

```diff
@@ -128,12 +142,20 @@ private Optional<PlanNodeCanonicalInfo> loadValue(Session session, CacheKey key,
                 if (loadValueTimeout(startTimeInNano, timeoutInMilliseconds)) {
                     return Optional.empty();
                 }
-                inputTableStatisticsBuilder.add(getPlanStatisticsForTable(session, scanNode, enableVerboseRuntimeStats));
+                PlanNodeCanonicalInfo.InputTableCacheKey inputTableCacheKey = getInputTableCacheKey(scanNode);
+                if (getHistoryBasedOptimizerInputStatisticsCheckStrategy(session).equals(ALWAYS)) {
+                    inputTableStatisticsBuilder.add(getPlanStatisticsForTable(session, inputTableCacheKey, enableVerboseRuntimeStats));
+                }
+                else {
+                    inputTableCacheKeyBuilder.add(inputTableCacheKey);
```

Comment:
Otherwise only cache the input table cache key, which can be used to fetch input table statistics after query completes and writing statistics back